src/main/java/org/veupathdb/service/eda/ss/model/reducer/StreamUnionMerger.java

```diff
@@ -0,0 +1,62 @@
+package org.veupathdb.service.eda.ss.model.reducer;
+
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.List;
+import java.util.PriorityQueue;
+
+/**
+ * Merges two or more sorted streams of ID indexes, outputting a sorted stream of sorted ID indexes which represents
+ * a union of the input streams.
+ *
+ * This is done by maintaining a min heap containing the next element from each input iterator. After removing an
+ * element from the heap, we add the next element from that stream.
+ */
+public class StreamUnionMerger implements Iterator<Long> {
+  // A possible optimization is to use a slimmed-down min-heap as opposed to a PriorityQueue (implemented as a heap).
+  private PriorityQueue<HeapElement> minHeap;
+  private long lastIdIndex;
+
+  public StreamUnionMerger(List<Iterator<Long>> streams) {
+    this.minHeap = new PriorityQueue<>(streams.size(), Comparator.comparing(element -> element.idIndex));
+    // Add the first element from each stream to the min heap
+    streams.forEach(stream -> {
+      if (stream.hasNext()) {
+        minHeap.add(new HeapElement(stream, stream.next()));
+      }
+    });
+  }
+
+  @Override
+  public boolean hasNext() {
+    return !minHeap.isEmpty();
+  }
+
+  @Override
+  public Long next() {
+    HeapElement next;
```

Comment:
Maybe add an `if (!hasNext()) throw new NoSuchElementException();`

Reply:
Or trust the contract; we are interested in speed and this is extra.

Reply:
I think it's a good idea to add for now. There will be a big fun profiling session in the near future and maybe we'll decide to remove it.